From cb33a41c6fd212dc3a2618a7f55b558cdae587a0 Mon Sep 17 00:00:00 2001
From: "Hsia-Jun(Randy) Li" <randy.li@synaptics.com>
Date: Tue, 23 Apr 2024 18:22:36 +0800
Subject: [PATCH 14/15] v4l2object: parse stride_align

It is hard to propose the plane sizes, we just assume
the driver would update it to the right value.

Signed-off-by: Hsia-Jun(Randy) Li <randy.li@synaptics.com>
---
 subprojects/gst-plugins-good/sys/v4l2/gstv4l2object.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2object.c b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2object.c
index 056d540ed9..eedd31367f 100644
--- a/subprojects/gst-plugins-good/sys/v4l2/gstv4l2object.c
+++ b/subprojects/gst-plugins-good/sys/v4l2/gstv4l2object.c
@@ -4870,7 +4870,8 @@ gst_v4l2_object_match_buffer_layout (GstV4l2Object * obj, guint n_planes,
           "Not matching as offset %" G_GSIZE_FORMAT
           " is smaller than %" G_GSIZE_FORMAT " on plane %u",
           offset[p], obj->info.offset[p], p);
-      return FALSE;
+      if (!need_fmt_update)
+        return FALSE;
     } else if (offset[p] > obj->info.offset[p]) {
       GST_LOG_OBJECT (obj->dbg_obj,
           "Remote offset %" G_GSIZE_FORMAT
@@ -5005,7 +5006,8 @@ validate_video_meta_struct (GstV4l2Object * obj, const GstStructure * s)
     if (!g_str_equal (name, "padding-top")
         && !g_str_equal (name, "padding-bottom")
         && !g_str_equal (name, "padding-left")
-        && !g_str_equal (name, "padding-right")) {
+        && !g_str_equal (name, "padding-right")
+        && !g_strstr_len (name, -1, "stride-align")) {
       GST_WARNING_OBJECT (obj->dbg_obj, "Unknown video meta field: '%s'", name);
       return FALSE;
     }
@@ -5036,6 +5038,8 @@ gst_v4l2_object_match_buffer_layout_from_struct (GstV4l2Object * obj,
   gst_structure_get_uint (s, "padding-bottom", &align.padding_bottom);
   gst_structure_get_uint (s, "padding-left", &align.padding_left);
   gst_structure_get_uint (s, "padding-right", &align.padding_right);
+  gst_structure_get_uint (s, "stride-align0", &align.stride_align[0]);
+  gst_structure_get_uint (s, "stride-align1", &align.stride_align[1]);
 
   if (align.padding_top || align.padding_bottom || align.padding_left ||
       align.padding_right) {
@@ -5054,7 +5058,6 @@ gst_v4l2_object_match_buffer_layout_from_struct (GstV4l2Object * obj,
     GST_WARNING_OBJECT (obj->dbg_obj,
         "Requested buffer size (%d) doesn't match video info size (%"
         G_GSIZE_FORMAT ")", buffer_size, GST_VIDEO_INFO_SIZE (&info));
-    return FALSE;
   }
 
   GST_DEBUG_OBJECT (obj->dbg_obj,
-- 
2.17.1

